File: src/jfan/fanx/util/FanUtil.java
//
// Copyright (c) 2008, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   4 Oct 08  Brian Frank  Creation
//
package fanx.util;

import java.util.*;
import fan.sys.*;

/**
 * FanUtil defines the mappings between the Fan and Java type systems.
 */
public class FanUtil
{

  /**
   * Convert Java class to Fan type.
   */
  public static Type toFanType(Class cls, boolean checked)
  {
    // TODO: optimize performance
    Type t = (Type)javaToFanTypes.get(cls.getName());
    if (t != null) return t;
    if (!checked) return null;
    throw Err.make("Not a Fan type: " + cls.getName()).val;
  }

  private static HashMap javaToFanTypes = new HashMap();
  static
  {
    if (Sys.ObjType == null) java.lang.Thread.dumpStack();
    javaToFanTypes.put("boolean",              Sys.BoolType);
    //javaToFanTypes.put("long",                 Sys.IntType);
    //javaToFanTypes.put("double",               Sys.FloatType);
    javaToFanTypes.put("java.lang.Object",     Sys.ObjType);
    javaToFanTypes.put("java.lang.Boolean",    Sys.BoolType);
    javaToFanTypes.put("java.lang.String",     Sys.StrType);
    javaToFanTypes.put("java.lang.Number",     Sys.NumType);
    javaToFanTypes.put("java.lang.Long",       Sys.IntType);
    javaToFanTypes.put("java.lang.Double",     Sys.FloatType);
    javaToFanTypes.put("java.math.BigDecimal", Sys.DecimalType);
  }

  /**
   * Return if the specified Java class represents an immutable type.
   */
  public static boolean isJavaImmutable(Class cls)
  {
    // TODO: optimize performance
    return javaImmutables.get(cls.getName()) != null;
  }

  private static HashMap javaImmutables = new HashMap();
  static
  {
    javaImmutables.put("java.lang.Boolean",    Boolean.TRUE);
    javaImmutables.put("java.lang.String",     Boolean.TRUE);
    javaImmutables.put("java.lang.Long",       Boolean.TRUE);
    javaImmutables.put("java.lang.Double",     Boolean.TRUE);
    javaImmutables.put("java.math.BigDecimal", Boolean.TRUE);
  }

  /**
   * Return if the Fan Type is represented as a Java class
   * such as sys::Int as java.lang.Long.
   */
  public static boolean isJavaRepresentation(Type t)
  {
    if (t.pod() != Sys.SysPod) return false;
    return t == Sys.ObjType   ||
           t == Sys.BoolType  ||
           t == Sys.StrType   ||
           t == Sys.IntType   ||
           t == Sys.FloatType ||
           t == Sys.NumType   ||
           t == Sys.DecimalType;
  }

  /**
   * Given a Fan qname, get the Java class name:
   *   sys::Obj  =>  java.lang.Object
   *   foo::Bar  =>  fan.foo.Bar
   */
  public static String toJavaClassName(String podName, String typeName)
  {
    if (podName.equals("sys"))
    {
      switch (typeName.charAt(0))
      {
        case 'B':
          if (typeName.equals("Bool")) return "java.lang.Boolean";
          break;
        case 'D':
          if (typeName.equals("Decimal")) return "java.math.BigDecimal";
          break;
        case 'F':
          if (typeName.equals("Float")) return "java.lang.Double";
          break;
        case 'I':
          if (typeName.equals("Int")) return "java.lang.Long";
          break;
        case 'N':
          if (typeName.equals("Num")) return "java.lang.Number";
          break;
        case 'O':
          if (typeName.equals("Obj")) return "java.lang.Object";
          break;
        case 'S':
          if (typeName.equals("Str")) return "java.lang.String";
          break;
      }
    }
    return "fan." + podName + "." + typeName;
  }

  /**
   * Given a Fan qname, get the Java implementation class name:
   *   sys::Obj   =>  fan.sys.FanObj
   *   sys::Float =>  fan.sys.FanFloat
   *   sys::Obj   =>  fan.sys.FanObj
   */
  public static String toJavaImplClassName(String podName, String typeName)
  {
    if (podName.equals("sys"))
    {
      switch (typeName.charAt(0))
      {
        case 'B':
          if (typeName.equals("Bool")) return "fan.sys.FanBool";
          break;
        case 'D':
          if (typeName.equals("Decimal")) return "fan.sys.FanDecimal";
          break;
        case 'F':
          if (typeName.equals("Float")) return "fan.sys.FanFloat";
          break;
        case 'I':
          if (typeName.equals("Int")) return "fan.sys.FanInt";
          break;
        case 'N':
          if (typeName.equals("Num")) return "fan.sys.FanNum";
          break;
        case 'O':
          if (typeName.equals("Obj")) return "fan.sys.FanObj";
          break;
        case 'S':
          if (typeName.equals("Str")) return "fan.sys.FanStr";
          break;
      }
    }
    return "fan." + podName + "." + typeName;
  }

  /**
   * Given a Fan qname, get the Java type signature:
   *   sys::Obj  =>  java/lang/Object
   *   foo::Bar  =>  fan/foo/Bar
   */
  public static String toJavaTypeSig(String podName, String typeName, boolean nullable)
  {
    if (podName.equals("sys"))
    {
      switch (typeName.charAt(0))
      {
        case 'B':
          if (typeName.equals("Bool"))
            return nullable ? "java/lang/Boolean" : "Z";
          break;
        case 'D':
          if (typeName.equals("Decimal")) return "java/math/BigDecimal";
          break;
        case 'F':
          if (typeName.equals("Float")) return "java/lang/Double";
          break;
        case 'I':
          if (typeName.equals("Int")) return "java/lang/Long";
          break;
        case 'N':
          if (typeName.equals("Num")) return "java/lang/Number";
          break;
        case 'O':
          if (typeName.equals("Obj")) return "java/lang/Object";
          break;
        case 'S':
          if (typeName.equals("Str")) return "java/lang/String";
          break;
        case 'V':
          if (typeName.equals("Void")) return "V";
          break;
      }
    }
    return "fan/" + podName + "/" + typeName;
  }

  /**
   * Given a Fan type, get the Java type signature:
   *   fan/sys/Duration
   */
  public static String toJavaTypeSig(Type t)
  {
    return toJavaTypeSig(t.pod().name(), t.name(), t.isNullable());
  }

  /**
   * Given a Fan type, get the Java member signature.
   *   Lfan/sys/Duration;
   */
  public static String toJavaMemberSig(Type t)
  {
    String sig = toJavaTypeSig(t);
    if (sig.length() == 1) return sig;
    return "L" + sig + ";";
  }

  /**
   * Given a Fan type, get its stack type: 'A', 'I', 'J', etc
   */
  public static int toJavaStackType(Type t)
  {
    if (!t.isNullable())
    {
      if (t == Sys.BoolType) return 'I';
      if (t == Sys.VoidType) return 'V';
    }
    return 'A';
  }

  /**
   * Given a Java type signature, return the implementation
   * class signature for methods and fields:
   *   java/lang/Object  =>  fan/sys/FanObj
   *   java/lang/Long    =>  fan/sys/FanInt
   * Anything returns itself.
   */
  public static String toJavaImplSig(String jsig)
  {
    if (jsig.length() == 1)
    {
      switch (jsig.charAt(0))
      {
        case 'Z': return "fan/sys/FanBool";
        default: throw new IllegalStateException(jsig);
      }
    }

    if (jsig.charAt(0) == 'j')
    {
      if (jsig.equals("java/lang/Object"))  return "fan/sys/FanObj";
      if (jsig.equals("java/lang/Boolean")) return "fan/sys/FanBool";
      if (jsig.equals("java/lang/String"))  return "fan/sys/FanStr";
      if (jsig.equals("java/lang/Long"))    return "fan/sys/FanInt";
      if (jsig.equals("java/lang/Double"))  return "fan/sys/FanFloat";
      if (jsig.equals("java/lang/Number"))  return "fan/sys/FanNum";
      if (jsig.equals("java/math/BigDecimal")) return "fan/sys/FanDecimal";
    }
    return jsig;
  }

}

File: src/nfan/fanx/util/FanUtil.cs
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   27 Sep 06  Andy Frank  Creation
//

using System;
using System.Collections;
using System.Text;

namespace Fanx.Util
{
  /// <summary>
  /// Utilites for dealing with type names in .NET.
  /// </summary>
  public class FanUtil
  {
    /// <summary>
    /// Convert .NET type to Fan type.
    /// </summary>
    public static Fan.Sys.Type toFanType(Type netType, bool check)
    {
      Fan.Sys.Type t = (Fan.Sys.Type)netToFanTypes[netType.FullName];
      if (t != null) return t;
      if (!check) return null;
      throw Fan.Sys.Err.make("Not a Fan type: " + netType).val;
    }

    /// <summary>
    /// Return if the specified .NET type represents an immutable type.
    /// </summary>
    public static bool isNetImmutable(Type netType)
    {
      return netImmutables[netType.FullName] != null;
    }

    /// <summary>
    /// Return if the Fan Type is represented as a .NET class
    /// such as sys::Int as Fan.Sys.Long.
    /// </summary>
    public static bool isNetRepresentation(Fan.Sys.Type t)
    {
      if (t.pod() != Fan.Sys.Sys.SysPod) return false;
      return t == Fan.Sys.Sys.ObjType   ||
             t == Fan.Sys.Sys.BoolType  ||
             t == Fan.Sys.Sys.IntType   ||
             t == Fan.Sys.Sys.FloatType ||
             t == Fan.Sys.Sys.NumType   ||
             t == Fan.Sys.Sys.DecimalType;
    }

    /// <summary>
    /// Return the .NET type name for this Fan pod and type.
    /// </summary>
    public static string toNetTypeName(Fan.Sys.Str podName, Fan.Sys.Str typeName)
    {
      return toNetTypeName(podName.val, typeName.val);
    }

    /// <summary>
    /// Return the .NET type name for this Fan pod and type.
    /// </summary>
    public static string toNetTypeName(string podName, string typeName)
    {
      if (podName == "sys")
      {
        switch (typeName[0])
        {
          case 'B':
            if (typeName == "Bool") return "Fan.Sys.Boolean";
            break;
          case 'D':
            if (typeName == "Decimal") return "Fan.Sys.BigDecimal";
            break;
          case 'F':
            if (typeName == "Float") return "Fan.Sys.Double";
            break;
          case 'I':
            if (typeName == "Int") return "Fan.Sys.Long";
            break;
          case 'N':
            if (typeName == "Num") return "Fan.Sys.Number";
            break;
          case 'O':
            if (typeName == "Obj") return "System.Object";
            break;
        }
      }
      return "Fan." + FanUtil.upper(podName, false) + "." + typeName;
    }

    /// <summary>
    /// Given a Fan qname, get the .NET implementation class name:
    /// </summary>
    public static string toNetImplTypeName(string podName, string typeName)
    {
      if (podName == "sys")
      {
        switch (typeName[0])
        {
          case 'B':
            if (typeName == "Bool") return "Fan.Sys.FanBool";
            break;
          case 'D':
            if (typeName == "Decimal") return "Fan.Sys.FanDecimal";
            break;
          case 'F':
            if (typeName == "Float") return "Fan.Sys.FanFloat";
            break;
          case 'I':
            if (typeName == "Int") return "Fan.Sys.FanInt";
            break;
          case 'N':
            if (typeName == "Num") return "Fan.Sys.FanNum";
            break;
          case 'O':
            if (typeName == "Obj") return "Fan.Sys.FanObj";
            break;
        }
      }
      return "Fan." + FanUtil.upper(podName, false) + "." + typeName;
    }

    /// <summary>
    /// Given a .NET type signature, return the implementation
    /// class signature for methods and fields:
    ///   System.Object   =>  Fan.Sys.FanObj
    ///   Fan.Sys.Double  =>  Fan.Sys.FanFloat
    /// Anything else returns itself.
    /// </summary>
    public static string toNetImplTypeName(string ntype)
    {
      if (ntype[0] == 'S')
      {
        if (ntype == "System.Object")  return "Fan.Sys.FanObj";
      }
      if (ntype[0] == 'F')
      {
        if (ntype == "Fan.Sys.Boolean") return "Fan.Sys.FanBool";
        if (ntype == "Fan.Sys.BigDecimal") return "Fan.Sys.FanDecimal";
        if (ntype == "Fan.Sys.Double") return "Fan.Sys.FanFloat";
        if (ntype == "Fan.Sys.Long") return "Fan.Sys.FanInt";
        if (ntype == "Fan.Sys.Number") return "Fan.Sys.FanNum";
      }
      return ntype;
    }

    /// <summary>
    /// Return the .NET method name for this Fan method name.
    /// </summary>
    public static string toNetMethodName(string fanName)
    {
      if (fanName == "equals") return "_equals";
      return fanName;
    }

    /// <summary>
    /// Return the Fan method name for this .NET method name.
    /// </summary>
    public static string toFanMethodName(string netName)
    {
      if (netName == "_equals") return "equals";
      return netName;
    }

    /// <summary>
    /// Return a new string, where the first letter is uppercase.
    /// If the string is a fully qualified type name, make each
    /// character following a '.' uppercase as well.
    /// </summary>
    public static string upper(string str) { return upper(str, true); }
    public static string upper(string str, bool recurse)
    {
      if (str.Length <= 1) return str.ToUpper();

      // Always first letter
      Char[] chars = str.ToCharArray();
      chars[0] = Char.ToUpper(chars[0]);

      if (recurse)
      {
        // Check for namespace
        for (int i=0; i<chars.Length-1; i++)
          if (chars[i] == '.')
            chars[i+1] = Char.ToUpper(chars[i+1]);
      }

      return new string(chars);
    }

    /// <summary>
    /// If the qualifed type name is a Fan type, then return
    /// the pod name the type should belong to, which should
    /// also be the assembly name.  If the type name does not
    /// appear to be a Fan type, return null.
    /// </summary>
    public static string getPodName(string qname)
    {
      if (!qname.StartsWith("Fan.")) return null;

      // Must have one more '.'
      int index = qname.IndexOf('.', 5);
      if (index == -1) return null;

      // Make first char lower case
      StringBuilder s = new StringBuilder(qname.Substring(4, index-4));
      s[0] = Char.ToLower(s[0]);
      return s.ToString();
    }

    /// <summary>
    /// Split a qualifed type name into two strings, where
    /// s[0] is the namespace, and s[1] is the simple type
    /// name.
    /// If qname represents a nested class, then s[0] will
    /// be the qualified type name of the parent, and s[1]
    /// will simply be the nested class name.
    /// </summary>
    public static string[] splitQName(string qname)
    {
      string[] s = new string[] { null, qname };
      int index = qname.LastIndexOf('/');
      if (index != -1)
      {
        s[0] = qname.Substring(0, index);
        s[1] = qname.Substring(index+1);
      }
      else
      {
        index = qname.LastIndexOf('.');
        if (index != -1)
        {
          s[0] = qname.Substring(0, index);
          s[1] = qname.Substring(index+1);
        }
      }
      return s;
    }

    private static Hashtable netToFanTypes = new Hashtable();
    private static Hashtable netImmutables = new Hashtable();

    static FanUtil()
    {
      netToFanTypes["System.Object"]      = Fan.Sys.Sys.ObjType;
      netToFanTypes["Fan.Sys.Boolean"]    = Fan.Sys.Sys.BoolType;
      netToFanTypes["Fan.Sys.BigDecimal"] = Fan.Sys.Sys.DecimalType;
      netToFanTypes["Fan.Sys.Double"]     = Fan.Sys.Sys.FloatType;
      netToFanTypes["Fan.Sys.Long"]       = Fan.Sys.Sys.IntType;
      netToFanTypes["Fan.Sys.Number"]     = Fan.Sys.Sys.NumType;

      netImmutables["Fan.Sys.Boolean"]    = true;
      netImmutables["Fan.Sys.BigDecimal"] = true;
      netImmutables["Fan.Sys.Double"]     = true;
      netImmutables["Fan.Sys.Long"]       = true;
      netImmutables["Fan.Sys.Number"]     = true;
    }

  }
}